src/JSInterop/Microsoft.JSInterop/test/JSRuntimeBaseTest.cs

```diff
@@ -283,14 +283,24 @@ public class BeginInvokeAsyncArgs
 
             public Func<Exception, string, string, object> OnDotNetException { get; set; }
 
-            protected override object OnDotNetInvocationException(Exception exception, string assemblyName, string methodName)
+            protected internal override void EndInvokeDotNet(string callId, bool success, object resultOrError, string assemblyName, string methodIdentifier, long dotNetObjectId)
             {
-                if (OnDotNetException != null)
+                if (OnDotNetException != null && !success)
+                {
+                    resultOrError = OnDotNetException(resultOrError as Exception, assemblyName, methodIdentifier);
+                }
+
+                if (!success && resultOrError is Exception ex)
+                {
+                    resultOrError = ex.ToString();
+                }
+                else if (!success && resultOrError is ExceptionDispatchInfo edi)
                 {
-                    return OnDotNetException(exception, assemblyName, methodName);
+                    resultOrError = edi.SourceException.ToString();
                 }
 
-                return base.OnDotNetInvocationException(exception, assemblyName, methodName);
+                var args = JsonSerializer.Serialize(new[] { callId, success, resultOrError }, new JsonSerializerOptions { PropertyNamingPolicy = JsonNamingPolicy.CamelCase });
+                BeginInvokeJS(0, "DotNet.jsCallDispatcher.endInvokeDotNetFromJS", args);
```

Comment:
Again since this is not part of the contract, I don't think the test class should do this. Can we update the unit tests to describe the contract as it now exists, not as it was before?